src/environmentd/tests/tracing.rs

```diff
@@ -30,37 +31,146 @@ async fn test_expected_spans() {
         ("peek_stage_finish", "SELECT 1"),
         ("peek_stage_explain_plan", "EXPLAIN SELECT 1"),
     ];
-    let asserts = tests
-        .iter()
-        .map(|(name, _)| {
-            (
-                *name,
-                registry
-                    .build()
-                    .with_name(*name)
-                    .was_closed_exactly(1)
-                    .finalize(),
-            )
-        })
-        .collect::<Vec<_>>();
 
     let server = test_util::TestHarness::default()
         .with_enable_tracing(true)
-        .with_fluent_assertions(registry)
+        .with_capture(storage.clone())
         .with_system_parameter_default("opentelemetry_filter".to_string(), "info".to_string())
         .start()
         .await;
+
     let client = server.connect().await.unwrap();
 
+    // Assert that there are no expected spans.
+    {
+        let storage = storage.lock();
+        for (name, _) in tests {
+            let stats = storage
+                .all_spans()
+                .filter_map(|span| (span.metadata().name() == *name).then(|| span.stats()))
+                .collect::<Vec<_>>();
+            assert!(stats.is_empty(), "{stats:?}");
+        }
+    }
+
     for (_, sql) in tests {
         client.batch_execute(sql).await.unwrap();
     }
-    for (name, assert) in asserts {
-        if !assert.try_assert() {
-            // Try one more time and print the name. Use .assert here because it prints a deeper
-            // cause than try_assert.
-            println!("assert failed for {name}");
-            assert.assert();
+
+    {
+        let storage = storage.lock();
+        for (name, _) in tests {
+            let stats = storage
+                .all_spans()
+                .filter_map(|span| (span.metadata().name() == *name).then(|| span.stats()))
+                .collect::<Vec<_>>();
+            let stat = stats.into_element();
+            // TODO: entered and exited can sometimes be > 1 (and so we can't assert == 1). Why does
+            // this happen? It's not bootstrapping, which we know because of the empty span assert
+            // above.
+            assert!(stat.entered > 0, "{name}: {stat:?}");
+            assert!(stat.exited > 0, "{name}: {stat:?}");
+            assert_eq!(stat.is_closed, true, "{name}: {stat:?}");
+        }
+    }
+}
+
+// Test that secrets are not leaked in tracing. Send many kinds of queries that should execute many
+// of the most common code paths. This is not guaranteed to exhaustively test all of our tracing.
```

Comment:
Seems like it would be nice to hook this up to something like SQLSmith so we can get a more thorough coverage. Definitely don't need to do that now but do you think it would be possible?

Reply:
I don't think that'd be a useful thing because sqlsmith's purpose is to find panics in the executor and optimizer paths, and secrets are not that. Randomized testing tools, although random, are each focused on finding a specific kind of bug, and that's not the problem we have here.